### PR TITLE
Incorrect Signature attach added

### PR DESCRIPTION
**Pull request by killshotrevival**
Support added for https://portswigger.net/kb/issues/00200900_jwt-signature-not-verified, incorrect signature verification attack.

### CODE REVIEW

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 1]
```diff
@@ -143,6 +143,31 @@ private boolean executePubliclyWellKnownHMacSecretAttack() {
...
+        clonedJWTHolder.setPayload(payloadJSONObject.toString());
+
+        if (this.serverSideAttack.getJwtActiveScanRule().isStop()) {
+            return false;
+        }
+
+        if (verifyJWTToken(cloneJWTHolder.getBase64EncodedToken(), serverSideAttack)) {
+            raiseAlert(
+                    MESSAGE_PREFIX,
+                    VulnerabilityType.NULL_BYTE,
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:161]: Please create a new [VulnerabilityType](https://github.com/SasanLabs/owasp-zap-jwt-addon/blob/cb0b85efa9c291e3a5eeb35b755e7fbc35c4f2bc/src/main/java/org/zaproxy/zap/extension/jwt/utils/VulnerabilityType.java#L20) and use it here.

Comment (preetkaran20) [reply src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:161]: Please also build the message to show in the alert like [Message](https://github.com/SasanLabs/owasp-zap-jwt-addon/blob/595e00ad997688c494bdfeb7819dd102619417f1/src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties#L72)

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 2]
```diff
@@ -362,7 +387,8 @@ public boolean executeAttack(ServerSideAttack serverSideAttack) {
             return this.executeCustomPrivateKeySignedJWTTokenAttack()
                     || this.executeAlgoKeyConfusionAttack()
                     || this.executeNullByteAttack()
-                    || this.executePubliclyWellKnownHMacSecretAttack();
+                    || this.executePubliclyWellKnownHMacSecretAttack()
+                    || this.executeIncorrectSignatureAttack();
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:391]: This attack should have more priority than other attacks, so can you please add this before `executeAlgoKeyConfusionAttack` ?

**Review by preetkaran20 — COMMENTED** [review]
Thank you adding this functionality. Few comments added and please fix the build failures

File: src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties [hunk 3]
```diff
@@ -79,6 +79,11 @@ jwt.scanner.server.vulnerability.signatureAttack.nullByte.desc=Signature bytes a
 jwt.scanner.server.vulnerability.signatureAttack.nullByte.refs=http://projects.webappsec.org/w/page/13246949/Null%20Byte%20Injection
 jwt.scanner.server.vulnerability.signatureAttack.nullByte.soln=Validate entire Signature and ensure the validating library or function does process beyond (or ignoring) null bytes.
 
+jwt.scanner.server.vulnerability.signatureAttack.incorrectSignature.name=JWT signature not verified
```
Comment (kingthorin) [inline src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties:82]: Should use title caps

File: src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties [hunk 4]
```diff
@@ -79,6 +79,11 @@ jwt.scanner.server.vulnerability.signatureAttack.nullByte.desc=Signature bytes a
...
+jwt.scanner.server.vulnerability.signatureAttack.incorrectSignature.desc=Some implementation fail to properly verify the signature of JWT token, which can result in them accepting tokens with invalid signatures.
```
Comment (kingthorin) [inline src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties:83]: I believe implementation should have an s making it plural. Same with “JWT token”.

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java (hunk 1 above)
Comment (killshotrevival) [reply src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java]: Updated @preetkaran20

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java (hunk 2 above)
Comment (killshotrevival) [reply src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java]: Updated

File: src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties (hunk 3 above)
Comment (killshotrevival) [reply src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties]: Noted, updating that @kingthorin

File: src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties (hunk 4 above)
Comment (killshotrevival) [reply src/main/resources/org/zaproxy/zap/extension/jwt/resources/Messages.properties]: Updated

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 5]
```diff
@@ -143,6 +143,34 @@ private boolean executePubliclyWellKnownHMacSecretAttack() {
...
 
+    /**
+     * Mis-matching the token signature and token data, to verify if the JWT implementation verifies the signature properly.
+     * A malicious user can exploit this vulnerability by supplying an arbitrary claim in the JWT payload to obtain 
+     * new privileges or impersonate other users
+     */
+    private boolean executeIncorrectSignatureAttack() {
+        JWTHolder cloneJWTHolder = new JWTHolder(this.serverSideAttack.getJwtHolder());
+        JSONObject payloadJSONObject = new JSONObject(cloneJWTHolder.getPayload());
+        payloadJSONObject.put("username", "admin");
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:154]: Please add username and admin as string constants.

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 6]
```diff
@@ -143,6 +143,34 @@ private boolean executePubliclyWellKnownHMacSecretAttack() {
...
     }
 
+    /**
+     * Mis-matching the token signature and token data, to verify if the JWT implementation verifies the signature properly.
+     * A malicious user can exploit this vulnerability by supplying an arbitrary claim in the JWT payload to obtain 
+     * new privileges or impersonate other users
+     */
+    private boolean executeIncorrectSignatureAttack() {
+        JWTHolder cloneJWTHolder = new JWTHolder(this.serverSideAttack.getJwtHolder());
+        JSONObject payloadJSONObject = new JSONObject(cloneJWTHolder.getPayload());
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:153]: please add try catch over this functionality to eat/log the json exception as payload can be non-json as well and we don't want to stop other attacks if there is any exception while parsing the payload.

**Review by preetkaran20 — APPROVED** [review]

**Review by preetkaran20 — COMMENTED** [review]
Also please add this functionality in https://github.com/SasanLabs/owasp-zap-jwt-addon/blob/master/CHANGELOG.md file.

Create a new release 1.0.3 and add this as well as another unreleased change under 1.0.3 release so that we can include this change in newer release.

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java (hunk 5 above)
Comment (killshotrevival) [reply src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java]: Updated @preetkaran20

**Comment by killshotrevival** [comment]
[CHANGELOG.md](https://github.com/SasanLabs/owasp-zap-jwt-addon/blob/master/CHANGELOG.md) updated with new 1.0.3 release

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,7 +4,11 @@ All notable changes to this add-on will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
 
 ### Unreleased
-- Ensure i18n resources are always initialized.
+
+
+### [1.0.3] - 2023-01-01
```
Comment (kingthorin) [inline CHANGELOG.md:9]: I believe this should be left inside unreleased, it’s updated as part of the release process I think.

Right, @preetkaran20 ?

Comment (kingthorin) [reply CHANGELOG.md:9]: Disregard.

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 8]
```diff
@@ -143,6 +146,43 @@ private boolean executePubliclyWellKnownHMacSecretAttack() {
...
+                        Alert.RISK_HIGH,
+                        Alert.CONFIDENCE_HIGH,
+                        cloneJWTHolder.getBase64EncodedToken(),
+                        serverSideAttack);
+                return true;
+            }
+            return false;
+        }
+        catch (JSONException ex) {
+            throw new JWTException(
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:180]: I think just logging the exception and returning false would be good in this case.

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 9]
```diff
@@ -76,6 +76,9 @@ public class SignatureAttack implements JWTAttack {
 
     private static final String MESSAGE_PREFIX =
             "jwt.scanner.server.vulnerability.signatureAttack.";
+    private static final String INCORRECT_SIGNATURE_KEY = "username";
+    private static final String INCORRECT_SIGNATURE_VALUE = "admin";
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:80]: These are not signature keys or values. we can rename them to USERNAME and ADMIN or USERNAME_KEY and USERNAME_VALUE or something appropriate.

**Review by preetkaran20 — COMMENTED** [review]
@killshotrevival Also the code is not compiling successfully. can you please import the required package and also test it.

**Comment by killshotrevival** [comment]
Code updated according to you suggestions @preetkaran20

File: src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java [hunk 10]
```diff
@@ -143,6 +147,42 @@ private boolean executePubliclyWellKnownHMacSecretAttack() {
...
+     * new privileges or impersonate other users
+     *
+     * @throws JWTException
+     */
+    private boolean executeIncorrectSignatureAttack() {
+        try {
+            JWTHolder cloneJWTHolder = new JWTHolder(this.serverSideAttack.getJwtHolder());
+            JSONObject payloadJSONObject = new JSONObject(cloneJWTHolder.getPayload());
+            payloadJSONObject.put(INCORRECT_SIGNATURE_PAYLOAD_KEY, INCORRECT_SIGNATURE_PAYLOAD_VALUE);
+            cloneJWTHolder.setPayload(payloadJSONObje_PAYLOADct.toString());
```
Comment (preetkaran20) [inline src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java:162]: @killshotrevival  a typo here?

Comment (preetkaran20) [reply src/main/java/org/zaproxy/zap/extension/jwt/attacks/SignatureAttack.java]: I corrected it.

**Comment by preetkaran20** [comment]
@killshotrevival LGTM, merging the PR. Thank you :)

**Pull request merged**